runtime/lua/vim/path.lua

```diff
@@ -0,0 +1,173 @@
+local M = {}
+
+--- Iterate over all the parents of the given file or directory.
+---
+--- Example:
+--- <pre>
+--- local root_dir
+--- for dir in vim.path.parents(vim.api.nvim_buf_get_name(0)) do
+---   if vim.fn.isdirectory(dir .. "/.git") == 1 then
+---     root_dir = dir
+---     break
+---   end
+--- end
+---
+--- if root_dir then
+---   print("Found git repository at", root_dir)
+--- end
+--- </pre>
+---
+---@param start (string) Initial file or directory.
+---@return (function) Iterator
+function M.parents(start)
+  local dir = start
+  return function()
+    local parent = M.dirname(dir)
+    if parent == dir then
+      return nil
+    end
+
+    dir = parent
+    return parent
+  end
+end
+
+--- Return the parent directory of the given file or directory
+---
+---@param file (string) File or directory
+---@return (string) Parent directory of {file}
+function M.dirname(file)
+  return vim.fn.fnamemodify(file, ':h')
+end
+
+--- Return the basename of the given file or directory
+---
+---@param file (string) File or directory
+---@return (string) Basename of {file}
+function M.basename(file)
+  return vim.fn.fnamemodify(file, ':t')
+end
+
+--- Return an iterator over the files and directories located in {path}
+---
+---@param path (string) Directory
+---@return Iterator over files and directories in {path}. Each iteration yields
+---        two values: name and type.
+function M.dir(path)
+  local fs = vim.loop.fs_scandir(path)
+  return function()
+    return vim.loop.fs_scandir_next(fs)
+  end
+end
+
+--- Find a file or directory in the given path.
+---
+--- Finds a file or directory {name} starting from {path}. If {upward} is "true"
+--- then the search traverses upward through parent directories; otherwise, the
+--- search traverses downward. Note that downward searches are recursive and may
+--- end up searching many directories! If {stop} is non-nil, then the search
+--- stops when the directory given in {stop} is reached.
+---
+---@param name (string) Name of the file or directory to find. Must be a base
+---            name, paths are not supported.
+---@param kwargs (table) Optional keyword arguments:
+---                       - path (string): Path to begin searching from. If
+---                              omitted, the current working directory is used.
+---                       - upward (boolean, default false): If true, search
+---                                upward through parent directories. Otherwise,
+---                                search through child directories
+---                                (recursively).
+---                       - stop (string): Stop searching when this directory is
```

Comment:
Will this stop _before_ or _after_ searching the specified directory? (Use case: do not search in `$HOME`.)

Reply:
Alternative: add a `skip (string|table)` option?